### PR TITLE
fix: Refactor how KAFKA_TOPICS is resolved to improve typing

### PR DESCRIPTION
**Pull request by untitaker**
KAFKA_TOPICS has nullable TopicDefinitions, but most users want a
non-nullable one.

The only reason it is nullable is so that outcomes-billing can fall back
to outcomes.

Therefore, add an indirection that implements this fallback in a central
position and returns a non-nullable TopicDefinition.

### CODE REVIEW

File: src/sentry/eventstream/base.py [hunk 1]
```diff
@@ -110,7 +110,7 @@ def _dispatch_post_process_group_task(
                 queue=queue,
             )
 
-    def _get_queue_for_post_process(self, event: Event) -> str:
```
Comment (untitaker) [inline src/sentry/eventstream/base.py:113]: Changes around `GroupEvent` are unrelated, that's just where the journey wrt typing started

File: src/sentry/conf/server.py [hunk 2]
```diff
@@ -3022,12 +3010,8 @@ def build_cdc_postgres_init_db_volume(settings: Any) -> dict[str, dict[str, str]
 }
 
 
-class TopicDefinition(TypedDict):
-    cluster: str
-
-
 # Cluster configuration for each Kafka topic by name.
-KAFKA_TOPICS: Mapping[str, Optional[TopicDefinition]] = {
+KAFKA_TOPICS: Mapping[str, Optional[sentry.utils.kafka_config.TopicDefinition]] = {
```
Comment (asottile-sentry) [inline src/sentry/conf/server.py:3014]: I've been trying to reduce the app imports in this file -- any imports in here tend to cause cycles which turns off type checking for settings annoyingly.  they're really tough to track down as well :(

File: src/sentry/utils/kafka_config.py [hunk 3]
```diff
@@ -91,3 +91,27 @@ def get_kafka_admin_cluster_options(
...
+
+
+def get_topic_definition(cluster: str) -> TopicDefinition:
+    defn = settings.KAFKA_TOPICS.get(cluster)
+    if defn is not None:
+        return defn
+    elif cluster == settings.KAFKA_OUTCOMES_BILLING:
+        return get_topic_definition(settings.KAFKA_OUTCOMES)
+    else:
+        assert False
```
Comment (asottile-sentry) [inline src/sentry/utils/kafka_config.py:117]: this should probably be `raise ValueError(f'unknown {cluster=}')`

File: src/sentry/utils/kafka_config.py [hunk 4]
```diff
@@ -91,3 +91,27 @@ def get_kafka_admin_cluster_options(
     return _get_kafka_cluster_options(
         cluster_name, ADMIN_SECTION, only_bootstrap=True, override_params=override_params
     )
+
+
+def _validate_topic_definitions():
```
Comment (asottile-sentry) [inline src/sentry/utils/kafka_config.py:96]:
```suggestion
def _validate_topic_definitions() -> None:
```

File: src/sentry/conf/server.py (hunk 2 above)
Comment (untitaker) [reply src/sentry/conf/server.py:3014]: this used to be a string-typed annotation `Optional["sentry.utils..."]`, but the formatter ripped it apart again

Isn't `from __future__ import annotations` supposed to deal with this anyway?

Comment (asottile-sentry) [reply src/sentry/conf/server.py:3014]: the problem is more that the type is defined outside the settings file than the quoting part

for some reason django-stubs cannot resolve cycles at all in the settings file -- even if they are legal

Comment (untitaker) [reply src/sentry/conf/server.py:3014]: should we have a types.py for types used in settings?

i need type definitions to be available in both the settings file and in app code

Comment (asottile-sentry) [reply src/sentry/conf/server.py:3014]: yeah that would probably work

File: src/sentry/utils/kafka_config.py (hunk 4 above)
Comment (untitaker) [reply src/sentry/utils/kafka_config.py]: done. we still have to do that? or shouldn't mypy infer that?

Comment (asottile-sentry) [reply src/sentry/utils/kafka_config.py]: a function with no signature is treated as untyped (`()-> Any`)

**Review by asottile-sentry — APPROVED** [review]

**Review by evanh — APPROVED** [review]

**Pull request merged**